frontend/js/index.js

```
@@ -1,282 +1,348 @@
-        const controlButton = event.target.closest('[id^="message-menu-"]');
+    /** сообщения */
+    const container = document.querySelector(".messages");
+    const indicator = document.getElementById("new-message-indicator");
 
-        if (controlButton) {
-            const messageId = controlButton.id.split('-')[2];
-            const menu = document.getElementById(`messageDropdown-${messageId}`);
-            const messageText = document.getElementById(`message-text-${messageId}`);
+    // Хранит информацию об открытых меню
+    let openMenuId = null;
+    let highlightedMessageId = null;
 
-            // Закрываем все меню и убираем выделение
-            document.querySelectorAll('.message-menu.show').forEach(menu => {
-                menu.classList.remove('show');
-            });
-            document.querySelectorAll('.message-text.highlight-border').forEach(el => {
-                el.classList.remove('highlight-border');
-            });
+    function renderMessages(messages, forceScroll = false) {
+        // Проверяем, что messages - это массив
+        if (!Array.isArray(messages)) {
+            console.error("renderMessages received invalid data:", messages);
+            return;
+        }
 
-            // Показываем только текущее меню и выделяем текст
-            if (menu && !menu.classList.contains('show')) {
-                menu.classList.add('show');
-                if (messageText) {
-                    messageText.classList.add('highlight-border');
+        const isScrolledToBottom =
+            container.scrollHeight - container.scrollTop <= container.clientHeight + 50;
+        const prevScrollHeight = container.scrollHeight;
+
+        // Сохраняем состояние открытого меню перед обновлением
+        if (!openMenuId) {
+            // Проверяем, есть ли открытое меню
+            const openMenuEl = document.querySelector('.message-menu.show');
+            if (openMenuEl) {
```

Очень уж ваш код любит в глубину расти. Fail fastу тут не хватает.

---

Ну и с точки зрения чисто моей вкусовщины. Я люблю явные проверки на тип. Вот тут я бы писал `openMenuEl !== null`. Потому что приведение типов оно непредсказуемое и всякие нули и пустые строки тоже не пройдут проверку, хотя могли бы. Старая привычка и когда вижу такой код немного начинают поднывать шрамы былых отладок 🥲